src/detect.h

```diff
@@ -1130,8 +1134,17 @@ typedef struct DetectEngineCtx_ {
     HashTable *non_pf_engine_names;
 
     const char *firewall_rule_file_exclusive;
+
+    /* user provided rate filter callbacks. */
+    SCDetectRateFilterFunc rate_filter_callback;
+
+    /* use provided data to be passed to rate_filter_callback. */
+    void *rate_filter_callback_arg;
```

Comment:
Just allow one callback for now, I'm not sure if a list of callbacks really makes sense here.

Reply:
We can expand this later?

Reply:
Yes, probably even in non-breaking ways if DetectEngineCtx is treated as an opaque data structure.  And of course a library user could just fan it out as needed.